apps/notifications/src/panel/state/create-listener-middleware.ts

```diff
@@ -0,0 +1,90 @@
+import type { MiddlewareAPI, AnyAction, Action, Middleware } from 'redux';
+
+type Handler = ( storeAPI: MiddlewareAPI, action: Action ) => void;
+
+interface AddListenersAction extends Action {
+	type: 'LISTENER-MIDDLEWARE/ADD_LISTENERS';
+	payload: Record< string, Handler[] >;
+}
+interface RemoveListenersAction extends Action {
+	type: 'LISTENER-MIDDLEWARE/REMOVE_LISTENERS';
+	payload: Record< string, Handler[] >;
+}
+interface ClearListenersAction extends Action {
+	type: 'LISTENER-MIDDLEWARE/CLEAR_LISTENERS';
+}
+type MiddlewareAction = AddListenersAction | RemoveListenersAction | ClearListenersAction;
+
+function createListenerMiddleware(): Middleware {
+	const listeners = new Map< string, Set< Handler > >();
+
+	return ( storeAPI: MiddlewareAPI ) =>
+		( next: ( action: Action ) => void ) =>
+		( anyAction: AnyAction ) => {
+			// We only want to notify listeners of actions that are objects.
+			// (That is, ignore thunks and other non-action objects.)
+			if ( typeof anyAction !== 'object' || typeof anyAction.type !== 'string' ) {
+				return next( anyAction );
+			}
+
+			const action = anyAction as MiddlewareAction;
+
+			switch ( action.type ) {
+				case 'LISTENER-MIDDLEWARE/ADD_LISTENERS': {
+					for ( const [ type, handlers ] of Object.entries( action.payload ) ) {
+						if ( ! listeners.has( type ) ) {
+							listeners.set( type, new Set< Handler >() );
+						}
+						handlers.forEach( ( handler ) => listeners.get( type )!.add( handler ) );
```

Comment:
You can easily avoid the `!` operator by always doing the map lookup just once: instead of `.has` + `.get`, do just one `.get`.

Reply:
I don't follow 🤔. Could you explain it further? 🙏

Reply:
The body of the loop can be rewritten as:
```js
let listenersForType = listeners.get( type );
if ( ! listenersForType ) {
  listenersForType = new Set< Handler >();
  listeners.set( type, listenersForType );
}
handlers.forEach( ( handler ) => listenersForType.add( handler ) );
```
This has two advantages:
- there's one `.get` call instead of `handlers.length + 1`
- TypeScript can figure all the types without any hints like `!`

Reply:
I just tried this and TypeScript still complained 😅. This seems to be a design limitation of Typescript and there's no perfect solution yet. (see https://github.com/microsoft/TypeScript/issues/9998).

I think it's fine to use `!` to silence the warnings here if we make sure it's always set.

Reply:
That's interesting because _my_ TypeScript doesn't complain and is happy 🙂 

<img width="535" alt="Screenshot 2024-08-26 at 11 22 41" src="https://github.com/user-attachments/assets/1a24d8b2-e645-4a04-a847-fc6d6d17038f">

Reply:
[Here's](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAEghgOwCYBsICcoF4oAoCU2AfFAG4D2AlkgNwBQAxuQgM7BQqVsQIYvZReAdygBZOGAA8bdJQQBzADRQAyhGCT4yNOiJEC9Jq3ahIAgOQAVAJoAFAKLnDzNlAAWiVHwBcsTzoBtAF0BYPo6NHZObl50FgAxcnRLcGgcaOAePgA6eXVcKFNofHpKADM8KABCKA4uTNiEpJTIQgBvOgBIDKy4xOTUgWFVdUk-bQwofRKunsbslnzC1MU5vn6W4voAXzoPCbjssqT7OAY3AoL9r0xCLBI1vubU7LgkJCv-ScISoA) the playground for the latest TypeScript version. Could you verify that this solution works on that too?

Reply:
Thanks, I see the error in the playground. What is special about the playground is that the `listenersForType` is a global top-level variable, and that it's referenced inside a callback function. Do any of these two things to fix it:
1. wrap the code in a `function foo() { ... }` so that `listenersForType` becomes a local variable. Error disappears.
2. Rewrite the loop to `for ( const handler of handlers )`. Error also disappears.

I think TypeScript covers the case when the callback is called async, and the possibility that the variable is later reassigned. Consider this:
```ts
let listeners = map.get( type ); // Set | undefined
if ( ! listeners ) { listeners = new Set() }; // just Set after this
handlers.then( ( handler ) => listeners.add( handler ) ); // The same code as .forEach but clearly async
/* some intermediate code */
listeners = undefined; // .then handler will fail if it runs after this
```
Here, because `listeners` is not scoped to a small block where all assignments are under control, TypeScript assumes that `listeners = undefined` can happen.

Reply:
That fixed it for me! Thanks for the detailed explanation! 🙇 

> Here, because `listeners` is not scoped to a small block where all assignments are under control, TypeScript assumes that `listeners = undefined` can happen.

I thought that the playground automatically assumed ESM mode, but apparently not 😆.